app/services/migration/failures_recorder.rb

```diff
@@ -0,0 +1,46 @@
+module Migration
+  class FailuresRecorder
```

Comment:
It doesn't feel quite right to have failure logic and cache references in the `DataMigration` class as well as the `FailureRecorder`. If we come to change how this is done in the future (writing them to a DB for example instead of the cache) I would only expect to have to change the logic in one place.

Would this class be better suited as a more generic module that both reads and writes, something like:

```
class FailureManager
  def initialize(data_manager)
  end
  
  def add_failure(failure)
    # add failure to cache
  end
  
  def failures
    # return all failures from cache
  end
end
```